### ISSUE TITLE
Report variable name reuse within the same function

### ISSUE THREAD

**Issue by bart2** [PyCQA/pylint#2322]
### Problem
Reusing variable names within the scope of a function should be discouraged, as it makes the code more confusing to read and potentially susceptible to bugs. See https://stackoverflow.com/a/9054628

### Solution
Given code like this:
```
#!/usr/bin/env python3
"""some module"""

def some_func():
    """some function"""

    some_var = 42
    print(some_var)

    some_var = 'reused'
    print(some_var)
```
`pylint` should report that `some_var` was reused.

**Comment by PCManticore** [comment]
Hi, thanks for submitting an issue!
This was also suggested in the past, but unfortunately it feels to be a check that's going to annoy everyone. We know we should not reuse variable names, but when a linter comes to us and complains about it, it suddenly gets too frustrating. And apart of this consideration, there's also the problem that sometimes there is no way knowing if the variable was reassigned to some other type or if it was just assigned to a type that `pylint` knows nothing about, resulting in potential confusing scenarios. As such, I think it's best not including this check in `pylint` right now, but it can definitely be a third party check that folks could install on their own!

**Comment by bart2** [comment]
I appreciate that you took time to consider this :) I was thinking about this a bit more and I guess `mypy` is a better tool to catch those kinds of issues.

**Comment by bart2** [comment]
Turns out there is already a check that covers redefinition of variable type:
```
pylint --load-plugins=pylint.extensions.redefined_variable_type ...
```
